### ISSUE TITLE
🐛 Bug: `scripts-name-casing` should ignore `pnpm:devPreinstall` (and maybe also hidden pnpm scripts starting with `.`)

### ISSUE THREAD

**Issue by andreww2012** [JoshuaKGoldberg/eslint-plugin-package-json#2047]
### Bug Report Checklist

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have [searched for related issues](https://github.com/michaelfaith/eslint-plugin-package-json/issues?q=is%3Aissue) and found none that matched my issue.

### Expected

Similar to #1368: `pnpm:devPreinstall` is the special [script only supported by pnpm](https://pnpm.io/scripts#pnpmdevpreinstall).

Additionally, starting with v11, pnpm also supports so called ["hidden scripts" which names start with a dot](https://pnpm.io/scripts#hidden-scripts).

Should we ignore such script names in regards to this rule?

I also kind of want to raise the not being able to ignore script names problem again, since who knows what other conventions package managers (i.e. possibly more niche ones like the recently born [nub](https://nubjs.com)) might come up with.

### Actual

The following script names get flagged by the reported rule and it's not possible to suppress that:

```json
{
  "pnpm:devPreinstall": "echo foo",
  ".bar": "echo bar"
}
```

### Version

1.6.2

### ESLint Version

10.8.0

### Additional Info

(just a sketch; happy to create a separate issue if we agree on the need of this option)

### Proposed `ignore` option

Should accept an array of strings. If a string starts with `/` and ends with `/<optional regex flag(s)>`, it's treated as a regex; otherwise as a plain string (on the edge case: this would be super weird to ignore a script name starting literally with `/`, but that would still be possible to ignore using the regex form).

The example of configuring the rule to ignore the reported script names:

```ts
const rules = {
  'package-json/scripts-name-casing': [
    2,
    {
      ignore: ['pnpm:devPreinstall', '/^\\.[a-z0-9]+(-[a-z0-9]+)*$/'],
    },
  ],
};
```

**Comment by michaelfaith** [comment]
Oh wow, good catch.  I actually wasn't aware of the new hidden script feature.  That's really cool.  I could see that being useful.  Yeah, definitely should fix both of those issues. Marking as Accepting PRs. 👍  If you don't end up taking it sooner, i should be able to get to it sometime this weekend.

> just a sketch; happy to create a separate issue if we agree on the need of this option

I think that's a good idea.  I'm generally in favor, but yeah would you mind opening a new issue for it, so we can iron out the specifics?  Appreciate it!

**Comment by michaelfaith** [comment]
For the hidden script use case, i could see us going one of two ways: 

- ignore the leading `.` but still enforce the casing requirement for the rest of the script name
- ignore it entirely if it has a leading `.`

I'm kind of leaning towards the latter.  Mainly because, the purpose of the script and bin name casing rules is to enforce a consistent format for them so that users having to type those script/bin names to execute them have a consistent experience.  But if hidden scripts aren't ones that users directly execute, should we hold those to the same standard?  What do you think?

**Comment by andreww2012** [comment]
> If you don't end up taking it sooner, i should be able to get to it sometime this weekend.

I guess I can work on this, but I only have 2 days left before the vacation, so I cannot guarantee I'd able to do this in the remaining time.


> For the hidden script use case, i could see us going one of two ways:
> 
>     * ignore the leading `.` but still enforce the casing requirement for the rest of the script name
> 
>     * ignore it entirely if it has a leading `.`

I'm slightly more in favor of the former because, despite dot-scripts are not supposed to be invoked directly, they still end up being typed in other scripts' commands. The analogy for me is prefixing JS variables with an underscore - in this case, no one is suddenly using, say, snake_case. That said, I get your point about indirect/hidden usage of such scripts - should we introduce an option, say, `ignoreDotNames?: boolean`, that would control whether kebab-case is not enforced for the hidden scripts, defaulting to `false`?

**Comment by michaelfaith** [comment]
> should we introduce an option, say, `ignoreDotNames?: boolean`, that would control whether kebab-case is not enforced for the hidden scripts, defaulting to `false`?

Once #2050 is implemented, I don't think we'd need another option just for those.  I'm ok with being consistent and continuing to enforce the casing on the hidden script names (after the `.`). 👍
